src/server/transaction/server/driver.go

```diff
@@ -225,60 +221,81 @@ func (d *driver) appendTransaction(
-
-		if err := d.txnEnv.WithReadContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
-			// Get the existing transaction and append the new requests
-			info := &transaction.TransactionInfo{}
-			err := d.transactions.ReadWrite(txnCtx.SqlTx).Get(txn.ID, info)
-			if err != nil {
-				return err
-			}
-
-			// Save the length so that we can check that nothing else modifies the
-			// transaction in the meantime
-			numRequests = len(info.Requests)
-			numResponses = len(info.Responses)
-			info.Requests = append(info.Requests, items...)
-			if newResponses != nil {
-				info.Responses = newResponses
-			}
+		info.Requests = append(info.Requests, items...)
+		return d.runTransaction(txnCtx, info)
+	})
+}
 
-			// This may error with a postgres-level transaction conflict, in which
-			// case we want to save the generated responses and reattempt.  The saved
-			// responses will contain any references to objects generated outside of
-			// the transaction and need to be reused on the next attempt.
-			info, err = d.runTransaction(txnCtx, info)
-			newResponses = info.Responses
+// updateTransaction accepts a function that uses and (possibly) updates a transaction and runs it
+// in either a Read- or WriteContext, taking care to save any new transaction info, except in the case
+// of an unretriable error
+func (d *driver) updateTransaction(
+	ctx context.Context,
+	writeTxn bool,
+	txn *transaction.Transaction,
+	f func(txnCtx *txncontext.TransactionContext, info *transaction.TransactionInfo) (*transaction.TransactionInfo, error)
+	) (*transaction.TransactionInfo, error) {
+	// Run this thing in a loop in case we get a conflict, time out after some tries
+	var newInfo transaction.TransactionInfo
+	attempt := func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		if err = d.transactions.ReadWrite(txnCtx.SqlTx).Get(txn.ID, &newInfo); err != nil {
+			return err
+		}
+		if newInfo, err = f(txnCtx, &newInfo); err != nil && col.IsErrTransactionConflict(err) {
+			// if we saw ErrTransactionConflict, assume something had to change the transaction response
+			// (such as a pipeline seeing an update outside the transaction), so return this error to
+			// escape the SQL transaction and run the storing code
+			return &transactionModifiedError{}
+		} else {
 			return err
-		}); err != nil {
-			return nil, err
 		}
+	}
 
-		info := &transaction.TransactionInfo{}
-		if err := col.NewSQLTx(ctx, d.db, func(sqlTx *sqlx.Tx) error {
-			// Update the existing transaction with the new requests/responses
-			return d.transactions.ReadWrite(sqlTx).Update(txn.ID, info, func() error {
-				if len(info.Requests) != numRequests || len(info.Responses) != numResponses {
-					// Someone else modified the transaction while we did the dry run
-					return &transactionConflictError{}
-				}
+	for i := 0; i < 10; i++ {
+		var err error
+		if writeTxn {
+			err = d.txnEnv.WithWriteContext(ctx, attempt)
```

Comment:
could also pass in one of these functions but the signatures are getting kind of cumbersome. I guess I could have defined some types for them